File: MyDailyBeat/API/VerveAPI/FXFormScreenNameCell.m
//
//  FXFormScreenNameCell.m
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 9/21/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//

#import "FXFormScreenNameCell.h"

@implementation FXFormScreenNameCell

- (void)setUp
{
    self.selectionStyle = UITableViewCellSelectionStyleNone;
    self.textLabel.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleRightMargin;
    self.textField.textAlignment = NSTextAlignmentRight;
    self.textField = [[UITextField alloc] initWithFrame:CGRectMake(0, 0, 200, 21)];
    self.textField.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin |UIViewAutoresizingFlexibleLeftMargin;
    self.textField.font = [UIFont systemFontOfSize:self.textLabel.font.pointSize];
    self.textField.textColor = [UIColor colorWithRed:0.275f green:0.376f blue:0.522f alpha:1.000f];
    self.textField.delegate = self;
    [self.contentView addSubview:self.textField];
    
    [self.contentView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self.textField action:NSSelectorFromString(@"becomeFirstResponder")]];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textDidChange) name:UITextFieldTextDidChangeNotification object:self.textField];
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    _textField.delegate = nil;
}

- (void)setValue:(id)value forKeyPath:(NSString *)keyPath
{
    //TODO: is there a less hacky fix for this?
    static NSDictionary *specialCases = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        specialCases = @{@"textField.autocapitalizationType": ^(UITextField *f, NSInteger v){ f.autocapitalizationType = v; },
                         @"textField.autocorrectionType": ^(UITextField *f, NSInteger v){ f.autocorrectionType = v; },
                         @"textField.spellCheckingType": ^(UITextField *f, NSInteger v){ f.spellCheckingType = v; },
                         @"textField.keyboardType": ^(UITextField *f, NSInteger v){ f.keyboardType = v; },
                         @"textField.keyboardAppearance": ^(UITextField *f, NSInteger v){ f.keyboardAppearance = v; },
                         @"textField.returnKeyType": ^(UITextField *f, NSInteger v){ f.returnKeyType = v; },
                         @"textField.enablesReturnKeyAutomatically": ^(UITextField *f, NSInteger v){ f.enablesReturnKeyAutomatically = !!v; },
                         @"textField.secureTextEntry": ^(UITextField *f, NSInteger v){ f.secureTextEntry = !!v; }};
    });
    
    void (^block)(UITextField *f, NSInteger v) = specialCases[keyPath];
    if (block)
    {
        if ([keyPath isEqualToString:@"textField.returnKeyType"])
        {
            //oh god, the hack, it burns
            self.returnKeyOverridden = YES;
        }
        
        block(self.textField, [value integerValue]);
    }
    else
    {
        [super setValue:value forKeyPath:keyPath];
    }
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    CGRect labelFrame = self.textLabel.frame;
    labelFrame.size.width = MIN(MAX([self.textLabel sizeThatFits:CGSizeZero].width, 97), 240);
    self.textLabel.frame = labelFrame;
    
    CGRect textFieldFrame = self.textField.frame;
    textFieldFrame.origin.x = self.textLabel.frame.origin.x + MAX(97, self.textLabel.frame.size.width) + 5;
    textFieldFrame.origin.y = (self.contentView.bounds.size.height - textFieldFrame.size.height) / 2;
    textFieldFrame.size.width = self.textField.superview.frame.size.width - textFieldFrame.origin.x - 10;
    if (![self.textLabel.text length])
    {
        textFieldFrame.origin.x = 10;
        textFieldFrame.size.width = self.contentView.bounds.size.width - 10 - 10;
    }
    else if (self.textField.textAlignment == NSTextAlignmentRight)
    {
        textFieldFrame.origin.x = self.textLabel.frame.origin.x + labelFrame.size.width + 5;
        textFieldFrame.size.width = self.textField.superview.frame.size.width - textFieldFrame.origin.x - 10;
    }
    self.textField.frame = textFieldFrame;
}

- (void)update
{
    self.textLabel.text = self.field.title;
    self.textField.placeholder = [self.field.placeholder fieldDescription];
    self.textField.text = [self.field fieldDescription];
    self.textField.autocorrectionType = UITextAutocorrectionTypeNo;
    self.textField.autocapitalizationType = UITextAutocapitalizationTypeNone;
    self.textField.keyboardType = UIKeyboardTypeDefault;
}

- (BOOL)textFieldShouldBeginEditing:(__unused UITextField *)textField
{
    //welcome to hacksville, population: you
    if (!self.returnKeyOverridden)
    {
        //get return key type
        UIReturnKeyType returnKeyType = UIReturnKeyDone;
        UITableViewCell <FXFormFieldCell> *nextCell = self.nextCell;
        if ([nextCell canBecomeFirstResponder])
        {
            returnKeyType = UIReturnKeyNext;
        }
        
        self.textField.returnKeyType = returnKeyType;
    }
    return YES;
}

- (void)textFieldDidBeginEditing:(__unused UITextField *)textField
{
    [self.textField selectAll:nil];
}

- (void)textDidChange
{
    [self updateFieldValue];
}

- (BOOL)textFieldShouldReturn:(__unused UITextField *)textField
{
    if (self.textField.returnKeyType == UIReturnKeyNext)
    {
        [self.nextCell becomeFirstResponder];
    }
    else
    {
        [self.textField resignFirstResponder];
    }
    return NO;
}

- (void)textFieldDidEndEditing:(__unused UITextField *)textField
{
    [self updateFieldValue];
    
    if (self.field.action) self.field.action(self);
}

- (void)updateFieldValue
{
    self.field.value = self.textField.text;
}

- (BOOL)canBecomeFirstResponder
{
    return YES;
}

- (BOOL)becomeFirstResponder
{
    return [self.textField becomeFirstResponder];
}

- (BOOL)resignFirstResponder
{
    return [self.textField resignFirstResponder];
}

@end

File: MDB_Swift/MyDailyBeat/API/src/FXFormScreenNameCell.swift
//  Converted with Swiftify v1.0.6242 - https://objectivec2swift.com/
//
//  FXFormScreenNameCell.swift
//  VerveAPI
//
//  Created by Virinchi Balabhadrapatruni on 9/21/16.
//  Copyright © 2016 eVerveCorp. All rights reserved.
//
import FXForms
class FXFormScreenNameCell: FXFormDefaultCell, UITextFieldDelegate {
    var textField: UITextField!
    var isReturnKeyOverridden: Bool = false


    override func setUp() {
        self.selectionStyle = .none
        self.textLabel?.autoresizingMask = [.flexibleWidth, .flexibleBottomMargin, .flexibleRightMargin]
        self.textField.textAlignment = .right
        self.textField = UITextField(frame: CGRect(x: CGFloat(0), y: CGFloat(0), width: CGFloat(200), height: CGFloat(21)))
        self.textField.autoresizingMask = [.flexibleWidth, .flexibleBottomMargin, .flexibleLeftMargin]
        self.textField.font = UIFont.systemFont(ofSize: CGFloat((self.textLabel?.font?.pointSize)!))
        self.textField.textColor = UIColor(red: CGFloat(0.275), green: CGFloat(0.376), blue: CGFloat(0.522), alpha: CGFloat(1.000))
        self.textField.delegate = self
        self.contentView.addSubview(self.textField)
        self.contentView.addGestureRecognizer(UITapGestureRecognizer(target: self.textField, action: NSSelectorFromString("becomeFirstResponder")))
        NotificationCenter.default.addObserver(self, selector: #selector(self.textDidChange), name: NSNotification.Name.UITextFieldTextDidChange, object: self.textField)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
        self.textField.delegate = nil
    }

    override func layoutSubviews() {
        super.layoutSubviews()
        var labelFrame: CGRect = (self.textLabel?.frame)!
        let size = self.textLabel?.sizeThatFits(CGSize.zero)
        let temp = max((size?.width)!, 97)
        labelFrame.size.width = min(temp, 240)
        self.textLabel?.frame = labelFrame
        var textFieldFrame: CGRect = self.textField.frame
        textFieldFrame.origin.x = (self.textLabel?.frame.origin.x)! + max(97, (self.textLabel?.frame.size.width)!) + 5
        textFieldFrame.origin.y = (self.contentView.bounds.size.height - textFieldFrame.size.height) / 2
        textFieldFrame.size.width = (self.textField.superview?.frame.size.width)! - textFieldFrame.origin.x - 10
        if self.textLabel?.text?.characters.count == 0 {
            textFieldFrame.origin.x = 10
            textFieldFrame.size.width = self.contentView.bounds.size.width - 10 - 10
        }
        else if self.textField.textAlignment == .right {
            textFieldFrame.origin.x = (self.textLabel?.frame.origin.x)! + (labelFrame.size.width) + 5
            textFieldFrame.size.width = (self.textField.superview?.frame.size.width)! - textFieldFrame.origin.x - 10
        }

        self.textField.frame = textFieldFrame
    }

    override func update() {
        self.textLabel?.text = self.field.title
        self.textField.placeholder = (self.field.placeholder as! NSObject).fieldDescription()
        self.textField.text = self.field.fieldDescription()
        self.textField.autocorrectionType = .no
        self.textField.autocapitalizationType = .none
        self.textField.keyboardType = .default
    }

    func textFieldShouldBeginEditing(_ textField: UITextField) -> Bool {
        //welcome to hacksville, population: you
        if !self.isReturnKeyOverridden {
                //get return key type
            var returnKeyType: UIReturnKeyType = .done
            let nextCell: FXFormBaseCell? = self.nextCell as? FXFormBaseCell
            if (nextCell?.canBecomeFirstResponder)! {
                returnKeyType = .next
            }
            self.textField.returnKeyType = returnKeyType
        }
        return true
    }

    func textFieldDidBeginEditing(_ textField: UITextField) {
        self.textField.selectAll(nil)
    }

    func textDidChange() {
        self.updateFieldValue()
    }

    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        if self.textField.returnKeyType == .next {
            self.nextCell.becomeFirstResponder()
        }
        else {
            self.textField.resignFirstResponder()
        }
        return false
    }

    func textFieldDidEndEditing(_ textField: UITextField) {
        self.updateFieldValue()
        if let action = self.field.action {
            action(self)
        }
    }

    func updateFieldValue() {
        self.field.value = self.textField.text
    }

    override func becomeFirstResponder() -> Bool {
        return self.textField.becomeFirstResponder()
    }

    override func resignFirstResponder() -> Bool {
        return self.textField.resignFirstResponder()
    }
}
